### PR TITLE
Better detection of CBOR

### LINKED ISSUE #7640: CBOR: Improve recognition of CBOR data format
Currently we only check if the first byte of the body is a `BYTE_OBJECT_INDEFINITE` to determine whether the content is CBOR or not.  However, what we should actually do is to check whether the "major type" is an object.

See:

* https://github.com/FasterXML/jackson-dataformat-cbor/blob/master/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORParser.java#L614
* https://github.com/FasterXML/jackson-dataformat-cbor/blob/master/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORParser.java#L682

Also, CBOR can be prefixed with a self-identifying tag, `0xd9d9f7`,  which we should check for as well.  Currently Jackson doesn't recognise this tag, but it looks like that will change in the future: https://github.com/FasterXML/jackson-dataformat-cbor/issues/6

### PR DESCRIPTION
**Pull request by kimchy**
CBOR has a special header that is optional, if exists, allows for exact detection. Also, since we know which formats we support in ES, we can support the object major type case.
closes #7640

### CODE REVIEW

File: src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java [hunk 1]
```diff
@@ -208,14 +208,16 @@ public static XContentType xContentType(byte[] data) {
      * Guesses the content type based on the provided input stream.
      */
     public static XContentType xContentType(InputStream si) throws IOException {
-        int first = si.read();
-        if (first == -1) {
+        int iFirst = si.read();
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java]: can't you simply cast to byte directly? the signed extension is preserved though

File: src/test/java/org/elasticsearch/common/xcontent/XContentFactoryTests.java [hunk 2]
```diff
@@ -69,4 +71,21 @@ private void testGuessType(XContentType type) throws IOException {
             assertThat(XContentFactory.xContentType(builder.string()), equalTo(type));
         }
     }
+
+    public void testCBORBasedOnMajorObjectDetection() {
+        // for this {"foo"=> 5} perl encoder for example generates:
+        byte[] bytes = new byte[] {(byte) 0xA1, (byte) 0x43, (byte) 0x66, (byte) 6f, (byte) 6f, (byte) 5};
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/common/xcontent/XContentFactoryTests.java]: could we have more like those?

Comment (s1monw) [inline src/test/java/org/elasticsearch/common/xcontent/XContentFactoryTests.java]: maybe generate some examples with other langs and try to detect them?

**Comment by s1monw** [comment]
left some comments - this is also related to #10134

File: src/test/java/org/elasticsearch/common/xcontent/XContentFactoryTests.java (hunk 2 above)
Comment (mikemccand) [inline src/test/java/org/elasticsearch/common/xcontent/XContentFactoryTests.java]: Here's Python:

```
>>> ', '.join('(byte) 0x%x' % x for x in cbor.dumps({'foo': 5}))
'(byte) 0xa1, (byte) 0x63, (byte) 0x66, (byte) 0x6f, (byte) 0x6f, (byte) 0x5'
```

File: src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java (hunk 1 above)
Comment (kimchy) [inline src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java]: sure

File: src/test/java/org/elasticsearch/common/xcontent/XContentFactoryTests.java (hunk 2 above)
Comment (kimchy) [inline src/test/java/org/elasticsearch/common/xcontent/XContentFactoryTests.java]: thanks @mikemccand !

**Comment by kimchy** [comment]
@s1monw applied comments

**Comment by s1monw** [comment]
LGTM

**Comment by kimchy** [comment]
pushed

**Pull request closed without merging**